### DISCUSSION TITLE
Is Theia suitable to refactor (~rebuild) an old RCP Eclipse client application?

### DISCUSSION THREAD

**Discussion by gidthup** [eclipse-theia/theia#14591] [Q&A]
A customer is using an Eclipse RCP based client that was initially build around 2008-2010. The application is still in use but it is time for modernization, both client and server side The client is looking for refactoring, keeping much of the functionality the same from end-user perspective, also literally...'the Perspectives'.  The perspective functionality is a key requirement: Ability to manually reorganise views, layouts, stacking etc.  and save them as personal preferences.  Also dynamically adding new views programmatically, i.e. as result of a drill down action.      What I have seen sofar in the online demos of Theia is that a lot of this functionality is there, but is Theia suitable to further extend into an end-user application (with tabular column views, graph views, graphical (scada) views, etc. So functionality not specific to an IDE. 

Other frameworks like Angular, React, Vue dont support the perspective functionality that is available in Theia. There are some opensource 'grid' extensions, but that is not the same. 

So my question is..  Does anyone have experience using Theia for an application that is not an IDE?

Any suggestions or pointers would be very welcome.

**Comment by msujew** [comment]
Hey @gidthup,

> So my question is.. Does anyone have experience using Theia for an application that is not an IDE?

Personally, I would not recommend it. It's certainly possible, but if you don't need 95% of the feature set of Theia (i.e. all the IDE related parts), you're probably better off writing your own UI shell framework or built on top of stuff like [lumino](https://github.com/jupyterlab/lumino).

I would assume that you'd save time initially by building on top of Theia, but there are two major obstacles:

1. Existing abstractions in Theia: Once you need more custom-tailored behavior than what you get from the existing abstractions in Theia, you will need to rewrite the existing code anyway.
2. Breakage: Due to changing requirements, keeping up with VS Code and updated/outdated/exchanged dependencies we often have breaking changes in our releases. Building on top of Theia and trying to keep up to date with it is a relevant maintenance effort, that you likely don't want.

These, among others, will likely result in more work in the long run than just rolling your own UI app/framework.

**Reply by gidthup** [reply]
Thank you for the information Mark.   Potential breakage is something I didn't consider yet as we have been using a very stable eclipse RCP for many years.  The application is rather extensive (100+ screens) hence any backwards compatibility issues could result in a lot of rework.

**Reply by msujew** [reply]
Note that the breakage is something I might overstate. I've been involved in a project (see [here](https://github.com/arduino/arduino-ide)) that makes heavy use of the override feature available through our dependency injection mechanism. In this project, things break in unexpected ways during Theia upgrades due to certain behavioral changes that were done for specific (and rather unusual) requirements.

A lot of adopters don't try to modify existing behavior, but rather add new stuff on top - what you're probably looking at as well. In those cases, breakage is usually rather limited, since we aim to keep the existing contribution APIs pretty stable.

Also, I rather enjoy controlling the backbone of my application. It's likely why I mostly work on frameworks, and not end-user applications. This is also influencing my advice for rolling your own UI layer. So don't trust me completely ;)

There's also another issue that I like to point out with regards to deployability of Theia: Since its design is rooted in the IDE space, it is built with the expectation that there's one instance of the app per user. This is usually no problem if you're building a desktop Electron tool, but becomes a bit resource intensive if you plan to deploy a browser app. There are ways in theory to make Theia multi-tenant capable, but it's fairly complex.

**Comment by sdirix** [comment]
I don't believe Theia is exclusively suited for IDE-like applications. While Theia is undoubtedly designed with IDE use cases in mind, its modular architecture makes it a robust platform for building a wide variety of applications that aren't traditional IDEs. We've successfully built non-IDE applications using Theia, and here are some reasons why it makes sense:

- Theia provides out-of-the-box support for running applications in both browser and Electron environments, making it easy to build cross-platform solutions.
- Theia supports plugins and VS Code extensions, enabling you to reuse the vast ecosystem of existing VS Code extensions. You can also develop a custom extension mechanism using the same architecture, allowing 3-rd party extension support
- The Theia UI layer is designed to be minimal, allowing you to build custom, framework-agnostic UIs with only a thin integration layer for Theia.
- Features such as notifications, commands, keybindings, and preferences are broadly useful and can add value to a wide range of applications, even those that don’t resemble traditional IDEs.

*Note that Theia does not have a built-in perspective mechanism (in the sense of the Eclipse RCP), i.e. that is something which you have to develop on your own then. See also (https://github.com/eclipse-theia/theia/issues/14577)*

Therefore even if your application doesn’t look like a typical IDE, it can still benefit significantly from Theia’s rich feature set. However, it's important to recognize that not every application will benefit to the same degree. If most of these features are irrelevant to your use case, Theia might not be the best fit.

Using Theia comes with certain trade-offs:
- You’ll need to invest time in understanding the framework, its RPC mechanisms, and its reliance on Inversify.js for dependency injection.
- Theia’s core is relatively large, and even if you don’t need and hide some features (e.g., Monaco), they’ll still be included, which could add unnecessary overhead.

If you’re considering Theia, I recommend building a proof of concept (POC) to better understand its strengths and limitations. This will help you determine if it aligns with your needs.

Since you plan to write a lot of code from scratch, I suggest designing it to be largely agnostic of Theia (or any specific framework). By keeping your code decoupled and modular, you’ll retain flexibility:
- Allows you to reuse much of your code if you decide to switch frameworks later.
- Enables your code to run in various environments, such as frontend, backend, WebWorker, WebView, VS Code extensions, and more.

**Reply by gidthup** [reply]
Thank you Stefan. That are good points we have to consider. We plan to do a POC early next year, so this will help us to focus on subjects that will determine if Theia brings in our case.

**Comment by gschnepp** [comment]
After reading your answers here @msujew @sdirix, would you recommend migrating an RCP app to the web via either Theia or RAP? Some parts need to be rewritten anyway. Related question: Why use Theia with an Electron backend when the best use case for Theia seems to be nothing more than a more modern (bloated) desktop app that isn't really suitable for the web? I know this is a heretical question. I'm just expressing my uncertainty: when I started with RCP years ago, I thought RAP was the future-proof option for me, but it looks like Eclipse is now fully committed to Theia.

**Reply by msujew** [reply]
Hey @gidthup,

> Why use Theia with an Electron backend

There are quite a few reasons, but it boils down to three main points IMO:

1. Code/dependencies reuse from VS Code such as monaco, xterm, etc.
2. Native web stacks, which are just easier to build on (good luck trying to find devs who are proficient in RCP/RAP and looking for work)
3. Theia also supports the web use case with something like [theia-cloud](https://theia-cloud.io/). In fact, you can build a desktop and a web tool from the same code base.

>I thought RAP was the future-proof option for me, but it looks like Eclipse is now fully committed to Theia.

I'd argue no tech is future-proof. However, Theia is still a second class citizen on the [Eclipse download page](https://www.eclipse.org/downloads/), at least when compared to the Eclipse IDE.

**Reply by JonasHelming** [reply]
A few comments:

- I want to emphasis that Theia not just "supports" the web use case, it is actually natively built for it. It does not follow the "vanilla" web application architecture, though, as it has a real backend. This allows users to have a file system, launch processes, optional terminal access, etc. . These things are all necessary for building IDEs and many domain-specific tools. If you build a very classical business application which mainly directly operates on data bases and the user does not need a work space at all, it is not a good fit.

- RAP is still maintained and can be used to migrate RCP applications to the web with minimal changes. However, you should consider the alternative to directly migrate to a "native" web stack (with or without Theia). What RAP will not achieve for you is moving your tech stack to something that is used by Millions of web applications out there, e.g. React, Node, etc.

- I would also argue that no tech is future-proof at all. Technologies, especially open source ones rely on adoption, contributions and sponsorship. Even technologies backed by very large companies can kind of become obsolete over several months if strategies are adapted.

To answer your original question: It is impossible to provide a generic answer, as it depends on the size and type of your application. If you need to rewrite parts of it anyways, I would say probably Theia or a web tech stack would be what I would look at first.

What might make sense is to get help from people who have migrated other RCP applications before, for an initial judgement.

We (EclipseSource) and others provide services like this:
https://eclipsesource.com/technology/eclipse-rcp-technology/#migrating-eclipse-rcp-applications-to-the-web
